climada/util/api_client.py

```diff
@@ -0,0 +1,472 @@
+    """Checks sanity of downloaded file by comparing actual and registered check sum.
+
+    Parameters
+    ----------
+    local_path : Path
+        the downloaded file
+    filinfo : FileInfo
+        file information from CLIMADA data API
+
+    Raises
+    ------
+    Download.Failed
+        if the file is not what it's supposed to be
+    """
+    raise NotImplementedError("sanity check by hash sum needs to be implemented yet")
+
+
+class Client():
+    """Python wrapper around REST calls to the CLIMADA data API server.
+    """
+    MAX_WAITING_PERIOD = 6
+
+    class AmbiguousResult(Exception):
+        """Custom Exception for Non-Unique Query Result"""
+
+    class NoResult(Exception):
+        """Custom Exception for No Query Result"""
+
+    def __init__(self):
+        """Constructor of Client.
+
+        Data API host and chunk_size (for download) are configurable values.
+        Default values are 'climada.ethz.ch' and 8096 respectively.
+        """
+        self.headers = {"accept": "application/json"}
+        self.host = CONFIG.data_api.host.str().rstrip("/")
+        self.chunk_size = CONFIG.data_api.chunk_size.int()
+
+    @staticmethod
+    def _request_200(url, **kwargs):
+        """Helper method, triaging successfull and failing requests.
+
+        Returns
+        -------
+        dict
+            loaded from the json object of a successful request.
+
+        Raises
+        ------
+        NoResult
+            if the response status code is different from 200
+        """
+        page = requests.get(url, **kwargs)
+        if page.status_code == 200:
+            return json.loads(page.content.decode())
+        raise Client.NoResult(page.content.decode())
+
+    def get_datasets(self, data_type=None, name=None, version=None, properties=None,
+                     status='active'):
+        """Find all datasets matching the given parameters.
+
+        Parameters
+        ----------
+        data_type : str, optional
+            data_type of the dataset, e.g., 'litpop' or 'draught'
+        name : str, optional
+            the name of the dataset
+        version : str, optional
+            the version of the dataset
+        properties : dict, optional
+            search parameters for dataset properties, by default None
+        status : str, optional
+            valid values are 'preliminary', 'active', 'expired', and 'test_dataset',
+            by default 'active'
+
+        Returns
+        -------
+        list of DatasetInfo
+        """
+        url = f'{self.host}/rest/datasets'
+        params = {
+            'data_type': data_type,
+            'name': name,
+            'version': version,
+            'status': '' if status is None else status,
+        }
+        params.update(properties if properties else dict())
+        return [DatasetInfo.from_json(ds) for ds in Client._request_200(url, params=params)]
+
+    def get_dataset(self, data_type=None, name=None, version=None, properties=None,
```

Comment:
is this function really necessary? Can't get_datasets just do both?

Reply:
The default usage of the client is meant to go along these lines:
```
client = Client()
dataset_info = client.get_dataset(data_type='rainfall', country='CHE')
downloaded_files = client.download_dataset(dataset)
do_something_with_(downloaded_files )
```
Like this, `get_dataset` would raise an Exception if there were more then one datasets matching the search parameters.
So if no exception is raised, one can be sure that the search parameters are sufficient to return the one and only desired dataset.
If `get_datasets` is used instead, one has to check for ambiguity right afterwards, to prevent arbitrary behaviour.